### DISCUSSION TITLE
Co-op (TAG) gameplay idea

### DISCUSSION THREAD

**Discussion by synthgloom** [ppy/osu#20849] [Ideas]
[Disclaimer: I am not a programmer so I apologize if this concept is unrealistic. Please give feedback if possible!]

**Overview**
Currently in co-op mode, players switch turns whenever a new combo begins. I think this is a good mechanic that should be kept, but I propose an additional alternative - a feature I'll call "sub-difficulties". It uses "player swap triggers" (bookmarks with extra steps) to determine when to swap players.

**The pitch**
Mappers could place bookmarks (player swap triggers) in the editor whenever they want to swap player's turns. These bookmarks would be enabled only when co-op mode is selected in a multiplayer lobby, making it an "additive feature" (in other words, the map can be played normally in singleplayer, but also collaboratively in multiplayer). Once the player swap triggers are arranged in the desired order, the mapper saves them to a sub-difficulty. When viewing a mapset on the website, an icon could be displayed to indicate a map has sub-difficulties for multiplayer - here's an [example](https://i.imgur.com/iWMFd6g.png) (the plus icon).

Player swap triggers would only really have two parameters: position in the song (timecode), and player order (value from 1 to 4). This would dictate at which point in the map players swap, and which player gets control. Ideally, this should be easily visualized in the map editor (represented with red, blue, green and yellow bookmark lines to represent each player value). However, these player swap triggers should only be visible when a sub-difficulty is being edited, to reduce visual clutter on the timeline.

Importantly, mappers can choose whether the sub-difficulty is designed for 2, 3 or 4 players. This would give a lot of control to the mapper in designing the most natural gameplay possible, catering to every possible player composition. Every difficulty could have multiple sub-difficulties to accommodate the different number of players, and players could manually choose the sub-difficulty they want to play in the multiplayer lobby.

As a last resort (for example, if a 2-player lobby decides to play a 4-player sub-difficulty), an automated system would adapt these higher player sub-difficulties to the current lobby size. For example: Player 3 triggers get assigned to Player 1, Player 2 triggers to Player 4, and so on. The design of this system is beyond the scope of my post (I'm not a programmer, sorry).

**Why should this be in the game?**
In osu, players usually play against each other, but it can be incredibly fun to play together, too. In my opinion, this feature provides an intuitive but highly flexible building block for a new dimension of collaborative play. 

Traditionally, TAG maps are extremely difficult - they feature cross-screen jumps, excessive streams and players swap extremely fast, sometimes after just 2 notes. Personally, I think TAG maps have a huge potential for improvement. With my proposed feature, we could design maps with way more control over how difficulty is distributed across players, through smart placement of player swap triggers. Sub-difficulties introduce many possible gameplay combinations: players could have short or long turns, which could be distributed evenly or unevenly, and each turn could serve specific patterns to each player (jumps, streams, finger control, speed...).

Another advantage of this system is that even players who aren't skilled mappers can now intuitively add co-op functionality. They can easily add sub-difficulties to any map they want, even if it previously didn't have this functionality. And if they don't like existing sub-difficulties, it would be very easy to make a custom one tailored to that player's needs.

**Footnote**
I think Ranked co-op maps should definitely be a feature in osu!lazer. Who knows what the ranking process for maps will look like when osu!lazer is released, but assuming the ranking process stays the same, sub-difficulties and ranking criteria for them should be defined closely to how we already rank maps.

**Comment by peppy** [comment]
I originally wanted something like this in tag coop, and think at some point it'd be great to see it happen. But first we need to implement tag coop.
